### PR TITLE
fix(ar_batch): honor presence/frequency penalties in the batched AR sampler (closes #156)

### LINKED ISSUE #156: Completing #120: ar_batch pump (_BatchedARJob) never passes token_counts, so presence/frequency penalties no-op at all temperatures
Follow-up to #120 (my penalty PR, merged in 2.0.0). I think #120 is working exactly as written for the paths it touched — this is about a path it didn't, so I'm opening it as a question in case the `ar_batch` behaviour is intentional.

**Summary.** Under `--scheduler-mode ar_batch`, presence/frequency penalties are silently dropped at every temperature. The serial and MTP paths honour them correctly; only the batched-AR pump doesn't. This matters because `mtplx start opencode` auto-selects `ar_batch`, so the flagship agentic lane is the one where penalties go dead.

**Where (v2.0.2):**
- `mtplx/server/openai.py:1868` `_make_sampler(self, job: _BatchedARJob)`. The temp≤0 branch (`:1871`) returns `argmax`, and the fused fast-path (`:1873`, guarded by `not (presence or frequency)`) only runs when *no* penalty is set — so neither drops a user penalty.
- The actual drop is `:1891-1892` `sample_one` → `_sample_from_logits(logprobs[0], job.sampler, rng)`, called **without any completion token counts**. `grep -c token_counts mtplx/server/openai.py` → `0` — the symbol is absent from the whole file, so the penalty term has nothing to subtract against and no-ops at all temperatures.
- `mtplx/commands/public.py:218-219` `OPENCODE_FAIR_BATCHING_DEFAULTS = {"scheduler_mode": "ar_batch", ...}`, applied by `_apply_opencode_fair_defaults` (`:10983`, called at `:12087`) unless the user passes `--scheduler-mode serial`.

**Repro (<5 min):**
1. `mtplx start opencode ...` (or `mtplx serve --scheduler-mode ar_batch`) with `--default-presence-penalty 1.5`.
2. Send a prompt that loops a dense model (the #102 case). The penalty has no effect; identical looping to `--default-presence-penalty 0`.
3. Same launch with `--scheduler-mode serial` → penalty takes effect. 200 OK in every case, no error/warning.

**Expected:** penalties apply under `ar_batch` as they do under serial/MTP.
**Actual:** penalties silently ignored under `ar_batch` (and therefore in the opencode quickstart lane).

**Workaround today:** launch with `--scheduler-mode serial` (penalties honoured, at the cost of the batched pump's concurrency).

**Why this bites the exact users #120 was for:** #120 exists to give loop-prone Qwen3.6 an anti-repetition lever for long agentic sessions (#102). A user who follows that remedy and uses the opencode quickstart gets `ar_batch` auto-selected and their penalty dropped — reintroducing the #102 loop in the flagship lane with a 200 and no signal.

**Suggested fix:** thread per-sequence completion `token_counts` into the `_BatchedARJob` pump and pass them to `_sample_from_logits` in `sample_one`, mirroring what `generate_ar`/`generate_mtpk` already do. Happy to send the PR if you agree the batched pump should honour penalties.

### PR DESCRIPTION
**Pull request by mmmugh**
Closes #156. Completes #120.

## The bug

Under `--scheduler-mode ar_batch`, `presence_penalty` and `frequency_penalty` are a **no-op at every temperature**. `_BatchedARGenerationService._make_sampler` has two independent holes:

1. `temp <= 0` short-circuits to a bare `mx.argmax` that bypasses `_sample_from_logits` entirely, so penalties can never apply on the greedy path.
2. The numpy path calls `_sample_from_logits` **without** `token_counts`, so penalties no-op at `temp > 0` too.

`grep -c token_counts mtplx/server/openai.py` → `0` on `main`.

This matters because `ar_batch` is the lane agentic clients land in, and `--default-presence-penalty` / `--default-frequency-penalty` bake a penalty into every request — which then silently does nothing.

## The fix

Compute `penalties_active` once, route penalized greedy requests through `_sample_from_logits` (which already applies penalties in its `temp <= 0` branch), and thread per-sequence completion counts into it.

**The counts come from a closure-local `Counter`, not from `job.tokens`.** This is the subtle part, and the obvious implementation is wrong:

mlx-lm's `BatchGenerator` is a one-step-lookahead pipeline — the same `GenerationBatch._step()` that samples `t_k` returns `t_(k-1)` to the pump (`mlx_lm/generate.py:1327-1378`). So `job.tokens` — appended in `emit_token`, *after* `generator.next()` returns — is always **one token stale** at sampler-call time. Sourcing counts from it leaves the token just emitted unpenalized at its first opportunity, which is exactly the immediate repeat a loop-breaking penalty exists to suppress.

Each sampler closure is bound to one sequence and sees every token it generates, so a closure-local counter is lag-free by construction and independent of mlx-lm's pipeline depth. It is also O(1) per step, versus serial `generate_ar`'s O(n) `Counter(tokens)` rebuild.

## Scope

- Serial `generate_ar` and `generate_mtpk` are untouched.
- The fused mlx-lm fast path is preserved for unpenalized requests. Only penalized requests leave it — and penalized `temp > 0` requests already took the numpy path before this change, so the only new cost is on `temp <= 0` + penalty, which is 100% broken today.
- No cross-request bleed: mlx-lm compacts `samplers` in lockstep with `uids` (`filter(keep)`, `generate.py:1383-1397`), and each closure owns a private counter regardless.

## Tests

`tests/test_server_openai.py` gains a test that drives the **real `BatchGenerator`** through the **real pump append ordering** and asserts per-position equality with a serial `generate_ar` reference, parametrized over greedy+presence, greedy+frequency, and temp>0+presence — plus a control pinning the unpenalized fused argmax path.

The round-trip shape is load-bearing: a closure-level unit test passes against the stale-`job.tokens` implementation and cannot see defect (2).

- RED on unpatched `main` (3/3)
- RED on the naive `Counter(job.tokens)` fix (3/3) — it emits `[7, 7, ...]` where serial emits `[7, 3, ...]`
- GREEN on this fix
- Full suite: 1674 passed, 5 skipped (`test_penalties.py` + `test_penalty_request_wiring.py` green — serial/MTP unaffected)

## Verified on a real model

Qwen3.6-27B, `--scheduler-mode ar_batch`, `temperature=0`, prompt driving a degenerate loop, 400 max tokens:

| | `presence_penalty=0.0` | `presence_penalty=1.5` |
|---|---|---|
| most-repeated word | **`hello` × 119** | `i` × 17 |
| lexical diversity | 0.377 | 0.514 |

Unpenalized burns the full budget in a `hello hello hello…` loop; penalized never forms it. Server log confirms `scheduler_lane: "ar_batch"` for both.

## Follow-ups (not in this diff)

- Greedy + penalties could stay fused on-device via `apply_penalties_mlx` + `mx.argmax` instead of the per-step host sync. Correctness first; the per-position test would keep such a variant honest.
- Serial's Loop Guard `penalty_overlay` is still unwired in the `ar_batch` lane.

---

## Relationship to #157

@SuperMarioYL opened #157 for this same issue two days before this PR — that one came first, and I only found it afterwards. The two diffs agree exactly on the `penalties_active` refactor and on routing penalized greedy requests off the bare `mx.argmax` short-circuit.

They differ in **where the penalty counts come from**, and I believe that difference is load-bearing:

| | counts source | result |
|---|---|---|
| #157 | `Counter(job.tokens)` | penalties fire **one token late** |
| #164 | closure-local `Counter` | per-position identical to serial `generate_ar` |

`job.tokens` is appended by `emit_token` *after* `generator.next()` returns, but the sampler runs one step **ahead** inside `next()` — mlx-lm's `GenerationBatch._step()` samples `t_k` in the same call that returns `t_(k-1)` (`generate.py:1327/1367/1378`). So at sampler-call time `job.tokens` is missing the token generated immediately before, which is exactly the token a loop-breaking penalty most needs to see:

```
serial generate_ar (correct)      : [7, 3, 7, 7, 7, 7]
ar_batch w/ Counter(job.tokens)   : [7, 7, 3, 3, 7, 7]   <- #157
ar_batch w/ closure-local counter : [7, 3, 7, 7, 7, 7]   <- this PR
```

At position 1 the immediate repeat (`7, 7`) escapes unpenalized. Details and a standalone repro are in https://github.com/youssofal/MTPLX/pull/157#issuecomment-4971854889.

This is easy to miss — my first patch was byte-identical to #157 here, and it passed a sampler-closure unit test. It is only visible on a round trip through the real `BatchGenerator` with the pump's append-after-`next()` ordering, which is why the test in this PR is shaped that way.

**I'm happy to close this in favor of #157 if that change is applied there** — @SuperMarioYL got to the issue first. I care more about the correct version shipping than about which PR it ships in.

### CODE REVIEW

**Comment by youssofal** [comment]
Closing with full credit, and thank you for the complete package here: you reported #156, diagnosed the dead-sampler mechanism, explained precisely why Counter(job.tokens) is one step stale (the pump appends asynchronously, so a job-list read lags the sampler by a step or two, and your mlx-lm line citations were exactly right), and sent a correct fix.

We shipped our implementation rather than this branch only because it was already validated end to end on the release lineage when your PR landed: same design (sampler-owned counter), with the counter attached to the job object so a mid-generation sampler rebuild keeps penalty state, plus greedy penalties-then-argmax parity tests against the serial path. Functionally these converge. Yours would have fixed the issue.

You are credited in the 2.1.0 release notes for the report, the diagnosis, and the fix. Please keep them coming.
